### ISSUE TITLE
Add eslint

### ISSUE THREAD

**Issue by gcochard** [spectcl/spectcl#3]
- [x] come to consensus on an `.eslintrc` file

After consensus reached, add it to the repo and add automatic linting on `pre-push`

**Comment by ryanbmilbourne** [comment]
Waiting on this issue until `eslint` reaches their 1.0 milestone.

**Comment by gcochard** [comment]
`eslint` has closed all issues on their `1.0.0` milestone, so we can probably get started on this.

Here is a list of all rules, grouped into `On by default` and `Off by default`. If we want to *change* the default, we should check the box. Once completed, we will generate our `.eslintrc` file.

## Potential Errors

The following rules point out areas where you might have made mistakes.
### On by default
* [ ] comma-dangle - disallow or enforce trailing commas
* [ ] no-cond-assign - disallow assignment in conditional expressions
* [ ] no-console - disallow use of `console` in the node environment
* [ ] no-constant-condition - disallow use of constant expressions in conditions
* [ ] no-control-regex - disallow control characters in regular expressions
* [ ] no-debugger - disallow use of `debugger`
* [ ] no-dupe-args - disallow duplicate arguments in functions
* [ ] no-dupe-keys - disallow duplicate keys when creating object literals
* [ ] no-duplicate-case - disallow a duplicate case label.
* [ ] no-empty-character-class - disallow the use of empty character classes in regular expressions
* [ ] no-empty - disallow empty statements
* [ ] no-ex-assign - disallow assigning to the exception in a `catch` block
* [ ] no-extra-boolean-cast - disallow double-negation boolean casts in a boolean context
* [ ] no-extra-semi - disallow unnecessary semicolons
* [ ] no-func-assign - disallow overwriting functions written as function declarations
* [ ] no-inner-declarations - disallow function or variable declarations in nested blocks
* [ ] no-invalid-regexp - disallow invalid regular expression strings in the `RegExp` constructor
* [ ] no-irregular-whitespace - disallow irregular whitespace outside of strings and comments
* [ ] no-negated-in-lhs - disallow negation of the left operand of an `in` expression
* [ ] no-obj-calls - disallow the use of object properties of the global object (`Math` and `JSON`) as functions
* [ ] no-regex-spaces - disallow multiple spaces in a regular expression literal
* [ ] no-sparse-arrays - disallow sparse arrays
* [ ] no-unreachable - disallow unreachable statements after a return, throw, continue, or break statement
* [ ] use-isnan - disallow comparisons with the value `NaN`
* [ ] valid-typeof - Ensure that the results of typeof are compared against a valid string

### Off by default
* [ ] no-extra-parens - disallow unnecessary parentheses
* [ ] no-reserved-keys - disallow reserved words being used as object literal keys
* [ ] no-unexpected-multiline - Avoid code that looks like two expressions but is actually one
* [ ] valid-jsdoc - Ensure JSDoc comments are valid

## Best Practices

These are rules designed to prevent you from making mistakes. They either prescribe a better way of doing something or help you avoid footguns.

### On by default

* [ ] no-fallthrough - disallow fallthrough of `case` statements
* [ ] no-octal - disallow use of octal literals
* [ ] no-redeclare - disallow declaring the same variable more than once
### Off by default
* [ ] accessor-pairs - Enforces getter/setter pairs in objects
* [ ] block-scoped-var - treat `var` statements as if they were block scoped
* [ ] complexity - specify the maximum cyclomatic complexity allowed in a program
* [ ] consistent-return - require `return` statements to either always or never specify values
* [ ] curly - specify curly brace conventions for all control statements
* [ ] default-case - require `default` case in `switch` statements
* [ ] dot-notation - encourages use of dot notation whenever possible
* [ ] dot-location - enforces consistent newlines before or after dots
* [ ] eqeqeq - require the use of `===` and `!==`
* [ ] guard-for-in - make sure `for-in` loops have an `if` statement
* [ ] no-alert - disallow the use of `alert`, `confirm`, and `prompt`
* [ ] no-caller - disallow use of `arguments.caller` or `arguments.callee`
* [ ] no-div-regex - disallow division operators explicitly at beginning of regular expression
* [ ] no-else-return - disallow `else` after a `return` in an `if`
* [ ] no-empty-label - disallow use of labels for anything other than loops and switches
* [ ] no-eq-null - disallow comparisons to null without a type-checking operator
* [ ] no-eval - disallow use of `eval()`
* [ ] no-extend-native - disallow adding to native types
* [ ] no-extra-bind - disallow unnecessary function binding
* [ ] no-floating-decimal - disallow the use of leading or trailing decimal points in numeric literals
* [ ] no-implied-eval - disallow use of `eval()`-like methods
* [ ] no-iterator - disallow usage of `__iterator__` property
* [ ] no-labels - disallow use of labeled statements
* [ ] no-lone-blocks - disallow unnecessary nested blocks
* [ ] no-loop-func - disallow creation of functions within loops
* [ ] no-multi-spaces - disallow use of multiple spaces
* [ ] no-multi-str - disallow use of multiline strings
* [ ] no-native-reassign - disallow reassignments of native objects
* [ ] no-new-func - disallow use of new operator for `Function` object
* [ ] no-new-wrappers - disallows creating new instances of `String`,`Number`, and `Boolean`
* [ ] no-new - disallow use of new operator when not part of the assignment or comparison
* [ ] no-octal-escape - disallow use of octal escape sequences in string literals, such as `var foo = "Copyright \251";`
* [ ] no-param-reassign - disallow reassignment of function parameters
* [ ] no-process-env - disallow use of `process.env`
* [ ] no-proto - disallow usage of `__proto__` property
* [ ] no-return-assign - disallow use of assignment in `return` statement
* [ ] no-script-url - disallow use of javascript: urls.
* [ ] no-self-compare - disallow comparisons where both sides are exactly the same
* [ ] no-sequences - disallow use of comma operator
* [ ] no-throw-literal - restrict what can be thrown as an exception
* [ ] no-unused-expressions - disallow usage of expressions in statement position
* [ ] no-useless-call - disallow unnecessary `.call()` and `.apply()`
* [ ] no-void - disallow use of `void` operator
* [ ] no-warning-comments - disallow usage of configurable warning terms in comments - e.g. `TODO` or `FIXME`
* [ ] no-with - disallow use of the `with` statement
* [ ] radix - require use of the second argument for `parseInt()`
* [ ] vars-on-top - requires to declare all vars on top of their containing scope
* [ ] wrap-iife - require immediate function invocation to be wrapped in parentheses
* [ ] yoda - require or disallow Yoda conditions

## Strict Mode

These rules relate to using strict mode.

* [ ] strict - controls location of Use Strict Directives

## Variables

These rules have to do with variable declarations.

### On by default

* [ ] no-delete-var - disallow deletion of variables
* [ ] no-undef - disallow use of undeclared variables unless mentioned in a `/*global */` block
* [ ] no-unused-vars - disallow declaration of variables that are not used in the code

### Off by default

* [ ] init-declarations - enforce or disallow variable initializations at definition
* [ ] no-catch-shadow - disallow the catch clause parameter name being the same as a variable in the outer scope
* [ ] no-label-var - disallow labels that share a name with a variable
* [ ] no-shadow-restricted-names - disallow shadowing of names such as `arguments`
* [ ] no-shadow - disallow declaration of variables already declared in the outer scope
* [ ] no-undef-init - disallow use of undefined when initializing variables
* [ ] no-undefined - disallow use of `undefined` variable
* [ ] no-use-before-define - disallow use of variables before they are defined

## Node.js

These rules are specific to JavaScript running on Node.js.

* [ ] callback-return - enforce return after a callback
* [ ] handle-callback-err - enforces error handling in callbacks
* [ ] no-mixed-requires - disallow mixing regular variable and require declarations
* [ ] no-new-require - disallow use of new operator with the `require` function
* [ ] no-path-concat - disallow string concatenation with `__dirname` and `__filename`
* [ ] no-process-exit - disallow `process.exit()`
* [ ] no-restricted-modules - restrict usage of specified node modules
* [ ] no-sync - disallow use of synchronous methods

## Stylistic Issues

These rules are purely matters of style and are quite subjective.

### On by default

* [ ] no-mixed-spaces-and-tabs - disallow mixed spaces and tabs for indentation

### Off by default

* [ ] array-bracket-spacing - enforce spacing inside array brackets
* [ ] brace-style - enforce one true brace style
* [ ] camelcase - require camel case names
* [ ] comma-spacing - enforce spacing before and after comma
* [ ] comma-style - enforce one true comma style
* [ ] computed-property-spacing - require or disallow padding inside computed properties
* [ ] consistent-this - enforces consistent naming when capturing the current execution context
* [ ] eol-last - enforce newline at the end of file, with no multiple empty lines
* [ ] func-names - require function expressions to have a name
* [ ] func-style - enforces use of function declarations or expressions
* [ ] indent - this option sets a specific tab width for your code
* [ ] key-spacing - enforces spacing between keys and values in object literal properties
* [ ] lines-around-comment - enforces empty lines around comments
* [ ] linebreak-style - disallow mixed 'LF' and 'CRLF' as linebreaks
* [ ] max-nested-callbacks - specify the maximum depth callbacks can be nested
* [ ] new-cap - require a capital letter for constructors
* [ ] new-parens - disallow the omission of parentheses when invoking a constructor with no arguments
* [ ] newline-after-var - allow/disallow an empty newline after variable declarations
* [ ] no-array-constructor - disallow use of the `Array` constructor
* [ ] no-continue - disallow use of the `continue` statement
* [ ] no-inline-comments - disallow comments inline after code
* [ ] no-lonely-if - disallow if as the only statement in an else block
* [ ] no-multiple-empty-lines - disallow multiple empty lines
* [ ] no-nested-ternary - disallow nested ternary expressions
* [ ] no-new-object - disallow use of the `Object` constructor
* [ ] no-spaced-func - disallow space between function identifier and application
* [ ] no-ternary - disallow the use of ternary operators
* [ ] no-trailing-spaces - disallow trailing whitespace at the end of lines
* [ ] no-underscore-dangle - disallow dangling underscores in identifiers
* [ ] no-unneeded-ternary - disallow the use of `Boolean` literals in conditional expressions
* [ ] object-curly-spacing - require or disallow padding inside curly braces
* [ ] one-var - allow or disallow one variable declaration per function
* [ ] operator-assignment - require assignment operator shorthand where possible or prohibit it entirely
* [ ] operator-linebreak - enforce operators to be placed before or after line breaks
* [ ] padded-blocks - enforce padding within blocks
* [ ] quote-props - require quotes around object literal property names
* [ ] quotes - specify whether backticks, double or single quotes should be used
* [ ] semi-spacing - enforce spacing before and after semicolons
* [ ] semi - require or disallow use of semicolons instead of ASI
* [ ] sort-vars - sort variables within the same declaration block
* [ ] space-after-keywords - require a space after certain keywords
* [ ] space-before-blocks - require or disallow space before blocks
* [ ] space-before-function-paren - require or disallow space before function opening parenthesis
* [ ] space-in-parens - require or disallow spaces inside parentheses
* [ ] space-infix-ops - require spaces around operators
* [ ] space-return-throw-case - require a space after `return`, `throw`, and `case`
* [ ] space-unary-ops - require or disallow spaces before/after unary operators
* [ ] spaced-comment - require or disallow a space immediately following the `//` or `/*` in a comment
* [ ] wrap-regex - require regex literals to be wrapped in parentheses

## ECMAScript 6

These rules are only relevant to ES6 environments.

* [ ] arrow-parens - Require parens in arrow function arguments
* [ ] arrow-spacing - Require space before/after arrow functions arrow
* [ ] constructor-super - verify `super()` callings in constructors
* [ ] generator-star-spacing - enforce the spacing around the `*` in generator functions
* [ ] no-class-assign - disallow modifying variables of class declarations
* [ ] no-const-assign - disallow modifying variables that are declared using `const`
* [ ] no-this-before-super - disallow to use `this`/`super` before `super()` calling in constructors.
* [ ] no-var - require `let` or `const` instead of `var`
* [ ] object-shorthand - require method and property shorthand syntax for object literals
* [ ] prefer-const - suggest using of `const` declaration for variables that are never modified after declared
* [ ] prefer-spread - suggest using of the spread operator instead of `.apply()`.
* [ ] require-yield - disallow generator functions that does not have `yield`

## Legacy

The following rules are included for compatibility with [JSHint](http://jshint.com/) and [JSLint](http://jslint.com/). While the names of the rules may not match up with the JSHint/JSLint counterpart, the functionality is the same.

* [ ] max-depth - specify the maximum depth that blocks can be nested
* [ ] max-len - specify the maximum length of a line in your program
* [ ] max-params - limits the number of parameters that can be used in the function declaration.
* [ ] max-statements - specify the maximum number of statement allowed in a function
* [ ] no-bitwise - disallow use of bitwise operators
* [ ] no-plusplus - disallow use of unary operators, `++` and `--`

**Comment by gcochard** [comment]
I'll start with the changes I've already made.

```yaml
  "no-extra-parens": 2
  "no-unexpected-multiline": 2
  "valid-jsdoc": 2
  "no-fallthrough": 2
  "no-octal": 2
  "no-redeclare": 2
  "default-case": 2
  "dot-notation": 2
  "eqeqeq": 2
  "no-alert": 2
  "no-caller": 2
  "no-else-return": 2
  "no-empty-label": 2
  "no-eval": 2
  "no-extend-native": 2
  "no-floating-decimal": 2
  "no-implied-eval": 2
  "no-iterator": 2
  "no-lone-blocks": 2
  "no-loop-func": 2
  "no-multi-spaces": 2
  "no-multi-str": 2
  "no-native-reassign": 2
  "no-new-func": 2
  "no-new-wrappers": 2
  "no-new": 2
  "no-octal-escape": 2
  "no-proto": 2
  "no-return-assign": 2
  "no-self-compare": 2
  "no-throw-literal": 2
  "no-unused-expressions": 2
  "no-warning-comments": 2
  "no-with": 2
  "radix": 2
  "strict": [2, "global"]
  "no-delete-var": 2
  "no-undef": 2
  "no-unused-vars": 2
  "no-shadow-restricted-names": 2
  "no-shadow": 2
  "no-undef-init": 2
  "no-use-before-define": 2
  "callback-return": 2
  "handle-callback-err": 2
  "no-mixed-requires": 2
  "no-new-require": 2
  "no-path-concat": 2
  "no-process-exit": 2
  "no-sync": 2
  "camelcase": 2
  "consistent-this": [2, "self"]
  "indent": 2
  "linebreak-style": 2
  "new-cap": 2
  "new-parens": 2
  "no-array-constructor": 2
  "no-multiple-empty-lines": [2, {"max": 2}]
  "no-nested-ternary": 2
  "no-new-object": 2
  "no-trailing-spaces": 2
  "quotes": [2, "single", "avoid-escape"]
  "semi": [2, "never"]
  "space-return-throw-case": 2
  "spaced-comment": [2, "always", {"exceptions": ["="]}]

**Comment by ryanbmilbourne** [comment]
After reviewing the checked list above, I would suggest that we check `no-fallthrough` in the `Best Practices - On by Default` section.  One of the valid use-cases (badjoke eel) for a switch block is fallthrough.  I get that the rule is mainly to make sure that we break out of our cases, but a lint error for intended fallthrough would get annoying, especially if we implement `eslint` in our pre-commit workflow.  I checked it above.  Otherwise, I'm happy with your suggested modifications.

**Comment by gcochard** [comment]
How about setting it to warning rather than error, so that it is noted but will not prevent a commit?

**Comment by ryanbmilbourne** [comment]
That sounds like it'd be just fine, and would be more in keeping with `eslint`'s defaults.
